filesystems/blob/src/main/scala/cromwell/filesystems/blob/BlobPathBuilderFactory.scala

```diff
@@ -1,27 +1,120 @@
 
   override def withOptions(options: WorkflowOptions)(implicit as: ActorSystem, ec: ExecutionContext): Future[BlobPathBuilder] = {
     Future {
-      new BlobPathBuilder(new AzureSasCredential(sasToken), container, endpoint)
+      new BlobPathBuilder(blobTokenGenerator, container, endpoint)
+    }
+  }
+}
+
+sealed trait BlobTokenGenerator {
+  def getAccessToken: String
+}
+
+object BlobTokenGenerator {
+  def createBlobTokenGenerator(container: String, endpoint: String): BlobTokenGenerator = {
+    createBlobTokenGenerator(container, endpoint, None, None)
+  }
+  def createBlobTokenGenerator(container: String, endpoint: String, workspaceId: Option[String], workspaceManagerURL: Option[String]): BlobTokenGenerator = {
+     (container: String, endpoint: String, workspaceId, workspaceManagerURL) match {
+       case (container, endpoint, None, None) =>
+         NativeBlobTokenGenerator(container, endpoint)
+       case (container, endpoint, Some(workspaceId), Some(workspaceManagerURL)) =>
+         WSMBlobTokenGenerator(container, endpoint, workspaceId, workspaceManagerURL)
+       case _ =>
+         throw new Exception("Bad! >:(")
+     }
+  }
+}
+
+case class WSMBlobTokenGenerator(container: String, endpoint: String, workspaceId: String, workspaceManagerURL: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
+    throw new NotImplementedError
+  }
+}
+
+case class NativeBlobTokenGenerator(container: String, endpoint: String) extends BlobTokenGenerator {
+  def getAccessToken: String = {
+    val storageAccountName = BlobPathBuilder.parseStorageAccount(BlobPathBuilder.parseURI(endpoint)) match {
+      case Some(storageAccountName) => storageAccountName
+      case _ => throw new Exception("Storage account could not be parsed from endpoint")
+    }
+
+    val profile = new AzureProfile(AzureEnvironment.AZURE)
+    val azureCredential = new DefaultAzureCredentialBuilder()
+      .authorityHost(profile.getEnvironment.getActiveDirectoryEndpoint)
+      .build
+    val azure = AzureResourceManager.authenticate(azureCredential, profile).withDefaultSubscription
+
+    val storageAccounts = azure.storageAccounts()
+    val storageAccount = storageAccounts
+      .list()
+      .asScala
+      .find(_.name == storageAccountName)
+
+    val storageAccountKeys = storageAccount match {
+      case Some(value) => value.getKeys().asScala.map(_.value())
+      case _ => throw new Exception("Storage Account not found")
+    }
+
+    val storageAccountKey = storageAccountKeys.headOption match {
+      case Some(value) => value
+      case _ => throw new Exception("Storage Account has no keys")
     }
+
+    val keyCredential = new StorageSharedKeyCredential(
+      storageAccountName,
+      storageAccountKey
+    )
+    val blobServiceClient = new BlobServiceClientBuilder()
+      .credential(keyCredential)
+      .endpoint(endpoint)
+      .buildClient()
+
+    val accountSasPermission = new AccountSasPermission()
```

Comment:
I wonder if the permissions that follow should be parameterized somehow -- or will these pretty much be the same for all uses?

Reply:
This is a good question, the WSM mediated token I think would change the permissions based on the requester (a cromwell reader only having read permission or a runner having read and write.). I am not sure we need that level of control for a native instance? But its something we should discuss with the team. This would be a good opportunity to break the permissions stuff out into a separate method, if its something we think we need.